### PR TITLE
[Bugfix] V1 sampler: route top_k=1 through greedy path to match temperature=0

### LINKED ISSUE #5404: [Bug]:  topk=1 and temperature=0 cause different output in vllm
### 🐛 Describe the bug

When using different generation configurations, such as top_k=1 or temperature=0 (while keeping other settings unchanged), why do the generated results change? They should both correspond to a deterministic greedy decoding.
vllm 0.4.3

---
Supplement:

The main issue encountered here is that the results generated by setting the temperature coefficient to 0 or topk to 1 are different. I understand that due to operator optimization and the lack of conventional arithmetic properties in floating-point numbers, matrix operations have a certain randomness. However, the sampling process occurs after the hidden_state is generated, at which point no calculations are involved. Therefore, the sampling results of the two sampling parameters should be the same.

### PR DESCRIPTION
**Pull request by qflen**
Fixes #5404.

## Summary

`top_k=1` and `temperature=0` can pick different tokens on the same logits.
`apply_top_k_only` masks with strict `<`, so all tokens tied at the maximum
survive; the random path's `probs.div(q).argmax` (Gumbel-max) then picks
among them via FP noise, disagreeing with `argmax`. Tied top logits are
common in fp16/bf16, which is why users keep hitting this.

## Fix

Extend the per-row `torch.where` selector in `Sampler.sample` so rows with
`top_k == 1` take the greedy branch in addition to rows with
`temperature < _SAMPLING_EPS`. No new flag, no backend changes,
`top_p`-only rows untouched. `top_p < 1.0` interaction is benign:
`apply_top_k_only(k=1)` already filters to the tied-top set and `top_p`
keeps at least the argmax, so it can't filter further.

## Test plan

New `tests/v1/sample/test_greedy_topk1_invariance.py` covers fp32/fp16/bf16,
mixed batch (`temperature=0` row vs `top_k=1` row, identical logits),
`top_k=1 + top_p<1.0`, the `all_random + top_k=1` path, and a guard that
`top_k>1` rows still reach the random path.

```
pytest tests/v1/sample/test_greedy_topk1_invariance.py \
       tests/v1/sample/test_sampler.py \
       tests/v1/sample/test_topk_topp_sampler.py \
       tests/v1/sample/test_batched_count_greater_than.py
# 57 passed, 123 skipped (CUDA-only)
pre-commit run --files vllm/v1/sample/sampler.py tests/v1/sample/test_greedy_topk1_invariance.py
pre-commit run mypy-3.10 --files <same> --hook-stage manual
# all hooks pass
```

CUDA backends untouched and not exercised locally.

## Duplicate-work check

- `gh issue view 5404`: open, 32 comments, marked unstale.
- `gh pr list ... --search "5404 in:body" / "sampler determinism" / "topk temperature" / "v1 sampler greedy"`: no open PR.
- Closed prior attempts #12802 and #13312 targeted the V0 sampler (gone from current main).

## Out of scope

`vllm/v1/worker/gpu/sample/sampler.py` uses the triton `gumbel_sample`
kernel directly (no `torch.where`). It already short-circuits at
`temp == 0`, but `top_k=1`/`temp>0` still passes through the seeded
gumbel-noise path. Symmetric fix would either modify the kernel or zero
out temperature for `top_k=1` rows before the kernel call. Different
surface, flagged as a follow-up rather than bundled.

---

Assisted by Claude Code (Opus 4.7, Max Effort) for debugging, every line reviewed by me.

### CODE REVIEW

File: vllm/v1/sample/sampler.py [hunk 1]
```diff
@@ -243,7 +243,11 @@ def sample(
...
         assert not (sampling_metadata.all_greedy and sampling_metadata.all_random)
-        if sampling_metadata.all_random:
+        # top_k=1 rows must take the greedy branch even when temperature>0:
+        # apply_top_k_only keeps every token tied with the maximum and the
+        # downstream Gumbel-max then picks among the survivors via FP noise,
+        # diverging from argmax (vllm-project/vllm#5404).
+        if sampling_metadata.all_random and sampling_metadata.top_k is None:
             greedy_sampled = None
         else:
             greedy_sampled = self.greedy_sample(logits)
```
[the remark this answers is not in the document]
Comment (qflen) [reply vllm/v1/sample/sampler.py]: ~ Pushed fix at [e36302c](https://github.com/vllm-project/vllm/pull/41598/commits/e36302c287a90007a605ff4a5b878d8d9c8eab81)
